Sherlock/replicator.h

```diff
@@ -0,0 +1,296 @@
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::function<void()> done_callback)
+        : valid_(false),
+          done_callback_(done_callback),
+          external_stop_(false),
+          remote_stream_(remote_stream, [this]() { TerminateSubscription(); }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
+          TerminateSubscription();
+        }
+        thread_.join();
+      } else {
+        if (done_callback_) {
+          done_callback_();
+        }
+      }
+    }
+
+   private:
+    static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
+    RemoteSubscriberThread() = delete;
+    RemoteSubscriberThread(const RemoteSubscriberThread&) = delete;
+    RemoteSubscriberThread(RemoteSubscriberThread&&) = delete;
+    void operator=(const RemoteSubscriberThread&) = delete;
+    void operator=(RemoteSubscriberThread&&) = delete;
+
+    void Thread() {
+      ThreadImpl();
+      subscriber_thread_done_ = true;
+      if (done_callback_) {
+        done_callback_();
+      }
+    }
+
+    void ThreadImpl() {
+      const RemoteStream& bare_stream = remote_stream_.ObjectAccessorDespitePossiblyDestructing();
+      bool terminate_sent = false;
+      while (true) {
+        if (!terminate_sent && external_stop_) {
+          terminate_sent = true;
+          if (subscriber_.Terminate() != ss::TerminationResponse::Wait) {
+            return;
+          }
+        }
+        try {
+          bare_stream.CheckSchema();
+          HTTP(ChunkedGET(
+              bare_stream.GetURLToSubscribe(index_),
+              [this](const std::string& header, const std::string& value) { OnHeader(header, value); },
+              [this](const std::string& chunk_body) { OnChunk(chunk_body); },
+              [this]() {}));
+        } catch (StreamTerminatedBySubscriber&) {
+          break;
+        } catch (current::Exception&) {
+        }
+      }
+    }
+
+    void OnHeader(const std::string& header, const std::string& value) {
+      if (header == "X-Current-Stream-Subscription-Id") {
+        std::unique_lock<std::mutex> lock(mutex_);
+        subscription_id_ = value;
+        *subscription_established_.MutableScopedAccessor() = true;
+      }
+    }
+
+    void OnChunk(const std::string& chunk) {
+      if (external_stop_) {
+        return;
+      }
+
+      // Expected one entire entry in each chunk,
+      // won't work in case of partial or multiple entries per chunk
+      const auto split = current::strings::Split(chunk, '\t');
+      assert(split.size() == 2u);
+      const auto idxts = ParseJSON<idxts_t>(split[0]);
+      assert(idxts.index == index_);
+      auto entry = ParseJSON<TYPE_SUBSCRIBED_TO>(split[1]);
+      const idxts_t unused_idxts;
```

Comment:
Makes me wonder if it's default-initialized correctly -- I know it is, but looks unsafe. Why not make it a `const` member variable and initialize in constructor?